File: packages/mcp-web/src/services/ddgs-search.ts
import { spawn } from 'child_process';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

export interface DDGSResult {
  title: string;
  url: string;
  description: string;
  provider: string;
}

export interface DDGSResponse {
  success: boolean;
  query: string;
  count?: number;
  backend?: string;
  region?: string;
  results?: DDGSResult[];
  error?: string;
}

/**
 * Search using DDGS Python library (metasearch across multiple engines)
 *
 * @param query Search query
 * @param maxResults Maximum number of results (default: 10)
 * @param region Region code (default: wt-wt for worldwide)
 * @param backend Search backend: auto, duckduckgo, bing, brave, google, etc. (default: auto for parallel search)
 */
export async function ddgsSearch(
  query: string,
  maxResults: number = 10,
  region: string = 'wt-wt',
  backend: string = 'auto'
): Promise<DDGSResponse> {
  return new Promise((resolve, reject) => {
    const scriptPath = join(__dirname, '../scripts/ddgs_search.py');

    // Spawn Python process
    const python = spawn('python3', [
      scriptPath,
      query,
      maxResults.toString(),
      region,
      backend,
    ]);

    let stdout = '';
    let stderr = '';

    python.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    python.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    python.on('close', (code) => {
      if (code !== 0 && code !== 1) {
        // Unexpected error code
        reject(new Error(`DDGS process exited with code ${code}: ${stderr}`));
        return;
      }

      try {
        const response: DDGSResponse = JSON.parse(stdout);
        if (response.success) {
          resolve(response);
        } else {
          reject(new Error(response.error || 'DDGS search failed'));
        }
      } catch (error) {
        reject(new Error(`Failed to parse DDGS output: ${stdout}\nStderr: ${stderr}`));
      }
    });

    python.on('error', (error) => {
      reject(new Error(`Failed to spawn Python process: ${error.message}`));
    });

    // Set timeout (30 seconds)
    const timeout = setTimeout(() => {
      python.kill();
      reject(new Error('DDGS search timed out after 30 seconds'));
    }, 30000);

    python.on('close', () => {
      clearTimeout(timeout);
    });
  });
}

File: packages/mcp-web/src/scripts/ddgs_search.py
#!/usr/bin/env python3
"""
DDGS Search Wrapper
Provides a simple CLI interface to the DDGS metasearch library
"""
import sys
import json
from ddgs import DDGS

def search(query: str, max_results: int = 10, region: str = "wt-wt", backend: str = "auto"):
    """
    Search using DDGS metasearch library

    Args:
        query: Search query
        max_results: Maximum number of results
        region: Region code (e.g., us-en, uk-en, wt-wt)
        backend: Search backend (auto, duckduckgo, bing, brave, google, etc.)
    """
    try:
        # Initialize DDGS (creates new session each time)
        ddgs = DDGS()

        # Perform search
        # DDGS.text() searches multiple backends in parallel when backend="auto"
        results = ddgs.text(
            keywords=query,
            region=region,
            safesearch="off",
            max_results=max_results,
            backend=backend
        )

        # Convert generator to list
        results_list = list(results)

        # Return JSON response
        response = {
            "success": True,
            "query": query,
            "count": len(results_list),
            "backend": backend,
            "region": region,
            "results": [
                {
                    "title": r.get("title", ""),
                    "url": r.get("href", ""),
                    "description": r.get("body", ""),
                    "provider": r.get("source", backend)
                }
                for r in results_list
            ]
        }

        print(json.dumps(response))
        return 0

    except Exception as e:
        # Return error as JSON
        error_response = {
            "success": False,
            "error": str(e),
            "query": query
        }
        print(json.dumps(error_response))
        return 1

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"success": False, "error": "Usage: ddgs_search.py <query> [max_results] [region] [backend]"}))
        sys.exit(1)

    query = sys.argv[1]
    max_results = int(sys.argv[2]) if len(sys.argv) > 2 else 10
    region = sys.argv[3] if len(sys.argv) > 3 else "wt-wt"
    backend = sys.argv[4] if len(sys.argv) > 4 else "auto"

    sys.exit(search(query, max_results, region, backend))
